File: android/src/new/java/com/margelo/nitro/rive/HybridViewModelInstance.kt
package com.margelo.nitro.rive

import android.util.Log
import androidx.annotation.Keep
import app.rive.RiveViewModelInstanceException
import app.rive.ViewModelInstance
import app.rive.ViewModelInstanceSource
import app.rive.runtime.kotlin.core.ViewModel
import app.rive.core.CommandQueue
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.Promise
import java.util.concurrent.ConcurrentHashMap

@Keep
@DoNotStrip
class HybridViewModelInstance(
  internal val viewModelInstance: ViewModelInstance,
  private val riveWorker: CommandQueue,
  private val parentFile: HybridRiveFile,
  private val viewModelName: String? = null,
  private val _instanceName: String? = null,
  // False for wrappers that alias an instance owned elsewhere (e.g. the
  // view's bound instance returned by RiveView.getViewModelInstance()).
  private val ownsInstance: Boolean = true
) : HybridViewModelInstanceSpec() {
  companion object {
    private const val TAG = "HybridViewModelInstance"
  }

  override fun dispose() {
    if (ownsInstance) {
      runCatching { viewModelInstance.close() }
    }
    super.dispose()
  }

  // TODO: Workaround — rive-android experimental SDK doesn't expose ViewModelInstance.name.
  // Only works when caller knows the name (createInstanceByName). Falls back to "" otherwise.
  override val instanceName: String
    get() = _instanceName ?: ""

  override fun getPropertiesAsync(): Promise<Array<ViewModelPropertyInfo>> {
    // "No metadata" must not read as "no properties" — instances obtained
    // via nested paths, list items, or the view don't carry their ViewModel
    // name on this backend.
    val name = viewModelName ?: return Promise.rejected(
      RuntimeException(
        "getPropertiesAsync is unavailable for this instance: its ViewModel " +
          "metadata is unknown (nested/list/view-obtained instances). Query the " +
          "ViewModel it was created from instead."
      )
    )
    val file = parentFile.riveFile
      ?: return Promise.rejected(RuntimeException("The RiveFile backing this instance was disposed"))
    return Promise.async {
      file
        .getViewModelProperties(name)
        .map { prop ->
        ViewModelPropertyInfo(name = prop.name, type = mapPropertyType(prop.type))
      }.toTypedArray()
    }
  }

  override fun numberProperty(path: String) =
    HybridViewModelNumberProperty(viewModelInstance, path)

  override fun stringProperty(path: String) =
    HybridViewModelStringProperty(viewModelInstance, path)

  override fun booleanProperty(path: String) =
    HybridViewModelBooleanProperty(viewModelInstance, path, ::hasBooleanProperty)

  private val booleanPathCache = ConcurrentHashMap<String, Boolean>()

  // rive-android 11.10+ answers a boolean read of an unknown path with an
  // uninitialized byte as the jboolean, which CheckJNI turns into a process
  // abort in debuggable builds (rive-app/rive-android#470). Resolve the path
  // against ViewModel metadata first; null means the lookup itself failed and
  // the read proceeds unguarded.
  internal suspend fun hasBooleanProperty(path: String): Boolean? {
    booleanPathCache[path]?.let { return it }
    val file = parentFile.riveFile ?: return null
    val parentPath = path.substringBeforeLast('/', "")
    val leaf = path.substringAfterLast('/')
    val result = try {
      val vmName = if (parentPath.isEmpty()) {
        viewModelName ?: viewModelInstance.getViewModelName()
      } else {
        val parent = try {
          ViewModelInstance.create(file, ViewModelInstanceSource.Reference(viewModelInstance, parentPath))
        } catch (e: RiveViewModelInstanceException) {
          return false.also { booleanPathCache[path] = it }
        }
        try {
          parent.getViewModelName()
        } finally {
          parent.close()
        }
      }
      file.getViewModelProperties(vmName).any {
        it.name == leaf && it.type == ViewModel.PropertyDataType.BOOLEAN
      }
    } catch (e: Exception) {
      RiveLog.w(TAG, "Could not resolve boolean property '$path': ${e.message}")
      return null
    }
    booleanPathCache[path] = result
    return result
  }

  override fun colorProperty(path: String) =
    HybridViewModelColorProperty(viewModelInstance, path)

  override fun enumProperty(path: String) =
    HybridViewModelEnumProperty(viewModelInstance, path)

  override fun triggerProperty(path: String) =
    HybridViewModelTriggerProperty(viewModelInstance, path)

  override fun imageProperty(path: String): HybridViewModelImagePropertySpec? {
    return try {
      HybridViewModelImageProperty(viewModelInstance, path, riveWorker)
    } catch (e: Exception) {
      Log.e(TAG, "imageProperty failed for path '$path'", e)
      null
    }
  }

  override fun listProperty(path: String): HybridViewModelListPropertySpec? {
    return try {
      HybridViewModelListProperty(viewModelInstance, path, riveWorker, parentFile)
    } catch (e: Exception) {
      Log.e(TAG, "listProperty failed for path '$path'", e)
      null
    }
  }

  override fun artboardProperty(path: String): HybridViewModelArtboardPropertySpec? {
    return try {
      HybridViewModelArtboardProperty(viewModelInstance, path, parentFile)
    } catch (e: Exception) {
      Log.e(TAG, "artboardProperty failed for path '$path'", e)
      null
    }
  }

  private fun viewModelImpl(path: String): HybridViewModelInstanceSpec? {
    val file = parentFile.riveFile ?: return null
    val source = ViewModelInstanceSource.Reference(viewModelInstance, path)
    val childVmi = ViewModelInstance.fromFile(file, source)
    return HybridViewModelInstance(childVmi, riveWorker, parentFile)
  }

  // Deprecated: Use viewModelAsync instead
  override fun viewModel(path: String): HybridViewModelInstanceSpec? {
    DeprecationWarning.warn("viewModel", "viewModelAsync")
    return try {
      viewModelImpl(path)
    } catch (e: Exception) {
      RiveLog.e(TAG, "viewModel failed for path '$path': ${e.message}")
      null
    }
  }

  override fun viewModelAsync(path: String): Promise<HybridViewModelInstanceSpec?> {
    return Promise.async { viewModelImpl(path) }
  }

  override fun replaceViewModel(path: String, instance: HybridViewModelInstanceSpec) {
    Log.w(TAG, "replaceViewModel not yet supported in experimental API")
  }
}

File: ios/legacy/HybridViewModelInstance.swift
import NitroModules
import RiveRuntime

class HybridViewModelInstance: HybridViewModelInstanceSpec {
  let viewModelInstance: RiveDataBindingViewModel.Instance?

  init(viewModelInstance: RiveDataBindingViewModel.Instance) {
    self.viewModelInstance = viewModelInstance
  }

  var instanceName: String { MainThread.run { viewModelInstance?.name ?? "" } }

  func getPropertiesAsync() throws -> Promise<[ViewModelPropertyInfo]> {
    throw RuntimeError.error(withMessage: "getPropertiesAsync is not supported on the legacy backend")
  }

  func numberProperty(path: String) throws -> (any HybridViewModelNumberPropertySpec)? {
    MainThread.run { () -> (any HybridViewModelNumberPropertySpec)? in
      guard let property = viewModelInstance?.numberProperty(fromPath: path) else { return nil }
      return HybridViewModelNumberProperty(property: property)
    }
  }

  func stringProperty(path: String) throws -> (any HybridViewModelStringPropertySpec)? {
    MainThread.run { () -> (any HybridViewModelStringPropertySpec)? in
      guard let property = viewModelInstance?.stringProperty(fromPath: path) else { return nil }
      return HybridViewModelStringProperty(property: property)
    }
  }

  func booleanProperty(path: String) throws -> (any HybridViewModelBooleanPropertySpec)? {
    MainThread.run { () -> (any HybridViewModelBooleanPropertySpec)? in
      guard let property = viewModelInstance?.booleanProperty(fromPath: path) else { return nil }
      return HybridViewModelBooleanProperty(property: property)
    }
  }

  func colorProperty(path: String) throws -> (any HybridViewModelColorPropertySpec)? {
    MainThread.run { () -> (any HybridViewModelColorPropertySpec)? in
      guard let property = viewModelInstance?.colorProperty(fromPath: path) else { return nil }
      return HybridViewModelColorProperty(property: property)
    }
  }

  func enumProperty(path: String) throws -> (any HybridViewModelEnumPropertySpec)? {
    MainThread.run { () -> (any HybridViewModelEnumPropertySpec)? in
      guard let property = viewModelInstance?.enumProperty(fromPath: path) else { return nil }
      return HybridViewModelEnumProperty(property: property)
    }
  }

  func triggerProperty(path: String) throws -> (any HybridViewModelTriggerPropertySpec)? {
    MainThread.run { () -> (any HybridViewModelTriggerPropertySpec)? in
      guard let property = viewModelInstance?.triggerProperty(fromPath: path) else { return nil }
      return HybridViewModelTriggerProperty(property: property)
    }
  }

  func imageProperty(path: String) throws -> (any HybridViewModelImagePropertySpec)? {
    MainThread.run { () -> (any HybridViewModelImagePropertySpec)? in
      guard let property = viewModelInstance?.imageProperty(fromPath: path) else { return nil }
      return HybridViewModelImageProperty(property: property)
    }
  }

  func listProperty(path: String) throws -> (any HybridViewModelListPropertySpec)? {
    MainThread.run { () -> (any HybridViewModelListPropertySpec)? in
      guard let property = viewModelInstance?.listProperty(fromPath: path) else { return nil }
      return HybridViewModelListProperty(property: property)
    }
  }

  func artboardProperty(path: String) throws -> (any HybridViewModelArtboardPropertySpec)? {
    MainThread.run { () -> (any HybridViewModelArtboardPropertySpec)? in
      guard let property = viewModelInstance?.artboardProperty(fromPath: path) else { return nil }
      return HybridViewModelArtboardProperty(property: property)
    }
  }

  func viewModel(path: String) throws -> (any HybridViewModelInstanceSpec)? {
    MainThread.run { () -> (any HybridViewModelInstanceSpec)? in
      guard let instance = viewModelInstance?.viewModelInstanceProperty(fromPath: path) else { return nil }
      return HybridViewModelInstance(viewModelInstance: instance)
    }
  }

  func replaceViewModel(path: String, instance: any HybridViewModelInstanceSpec) throws {
    guard let hybridInstance = instance as? HybridViewModelInstance,
          let nativeInstance = hybridInstance.viewModelInstance else {
      throw RuntimeError.error(withMessage: "Invalid ViewModelInstance provided to replaceViewModel")
    }
    let success = MainThread.run {
      viewModelInstance?.setViewModelInstanceProperty(fromPath: path, to: nativeInstance) ?? false
    }
    if !success {
      throw RuntimeError.error(withMessage: "Failed to replace ViewModel at path: \(path)")
    }
  }

  func viewModelAsync(path: String) throws -> Promise<(any HybridViewModelInstanceSpec)?> {
    return Promise.onMain { try self.viewModel(path: path) }
  }
}
